File: ois/ois20/round2/ois11_round2_solutions/ip.pas
{ NOTE: it is recommended to use this even if you don't understand the following code. }

{ input data }
var
    A    : array[0..3] of longint = (0, 0, 0, 0);
    B    : array[0..3] of longint = (0, 0, 0, 0);
    line : string;
    i, j : longint;
    An, Bn : int64;

begin
{
    uncomment the following lines if you want to read/write from files
    assign(input,  'input.txt');  reset(input);
    assign(output, 'output.txt'); rewrite(output);
}

    readln(line);
    j := 0;
    for i := 1 to length(line) do
        if line[i] = '.' then
            inc(j)
        else
            A[j] := A[j] * 10 + ord(line[i]) - ord('0');


    readln(line);
    j := 0;
    for i := 1 to length(line) do
        if line[i] = '.' then
            inc(j)
        else
            B[j] := B[j] * 10 + ord(line[i]) - ord('0');
    
    { at this point, the arrays A and B contain four elements each with the individual IP octects }
    { for example: 192.168.1.1 creates an array (192, 168, 1, 1) with 192 at index 0 }
    
    { insert your code here }
    An := 0;
    for i := 0 to 3 do An := 256*An + A[i];
    Bn := 0;
    for i := 0 to 3 do Bn := 256*Bn + B[i];

    writeln(Bn-An+1); { print result }
end.

File: ois/ois20/round2/ois11_round2_solutions/ip.py
#!/usr/bin/env python3
# NOTE: it is recommended to use this even if you don't understand the following code.


# input data
A = list(map(int, input().split(".")))
B = list(map(int, input().split(".")))

# at this point, the arrays A and B contain four elements each with the individual IP octects
# for example: 192.168.1.1 creates an array [192, 168, 1, 1] with 192 at index 0

ans = 0
exp = 0
for i in range(3, -1, -1):
    diff = B[i] - A[i]
    # Note: this works even when diff is less than 0
    ans += diff * (256 ** exp)
    if exp == 0:
        ans += 1
    exp += 1

print(ans)
